src/wasm/wasm-s-parser.cpp

```diff
@@ -845,7 +845,15 @@ void SExpressionWasmBuilder::preParseHeapTypes(Element& module) {
   types = builder.build();
 
   for (auto& pair : typeIndices) {
-    wasm.typeNames[types[pair.second]].name = pair.first;
+    auto name = pair.first;
+    auto type = types[pair.second];
+    // A type may appear in the type section more than once, but we canonicalize
+    // types internally, so there will be a single name chosen for that type. Do
+    // so determistically.
+    if (wasm.typeNames.count(type) && wasm.typeNames[type].name.str < name) {
+      continue;
+    }
+    wasm.typeNames[type].name = name;
```

Comment:
Why take the "smallest" name for a type rather than e.g. the first or last name that appears in the type section?

Reply:
No super-strong reasons, but: The smallest leads to the smallest output, which can be more readable. It also has the benefit that reordering the wast does not alter the result, which is intuitive since in the wast they are named and the order shouldn't matter in principle.